app/uk/gov/hmrc/cgtpropertydisposals/repositories/upscan/UpscanRepository.scala

```diff
@@ -0,0 +1,152 @@
+
+  def update(
+    draftReturnId: DraftReturnId,
+    upscanReference: UpscanReference,
+    upscanUpload: UpscanUpload
+  ): EitherT[Future, Error, Unit]
+
+}
+
+@Singleton
+class DefaultUpscanRepository @Inject() (mongo: ReactiveMongoComponent, config: Configuration)(
+  implicit val ec: ExecutionContext
+) extends ReactiveRepository[UpscanUpload, BSONObjectID](
+      collectionName = "upscan",
+      mongo          = mongo.mongoConnector.db,
+      UpscanUpload.format,
+      ReactiveMongoFormats.objectIdFormats
+    )
+    with UpscanRepository
+    with CacheRepository[UpscanUpload] {
+
+  val cacheTtl: FiniteDuration = config.underlying.get[FiniteDuration]("mongodb.upscan.expiry-time").value
+
+  override def indexes: Seq[Index] = Seq(
+    Index(
+      key = Seq("draftReturnId" → IndexType.Ascending, "upscanUploadMeta.reference" -> IndexType.Ascending)
+//      name     = Some("upscan-cache-ttl"),
+//      unique   = true,
+//      dropDups = true
+    )
+  )
+
+  override def insert(
+    upscanUpload: UpscanUpload
+  ): EitherT[Future, Error, Unit] =
+    EitherT[Future, Error, Unit](
+      collection.insert
+        .one[UpscanUpload](upscanUpload)
+        .map[Either[Error, Unit]] { result: WriteResult =>
+          if (result.ok)
+            Right(())
+          else
+            Left(
+              Error(
+                s"could not store upscan upload meta :${result.writeErrors}"
+              )
+            )
+        }
+        .recover {
+          case exception => Left(Error(exception))
+        }
+    )
+
+  override def select(
+    draftReturnId: DraftReturnId,
+    upscanReference: UpscanReference
+  ): EitherT[Future, Error, Option[UpscanUpload]] = {
+    val selector = Json.obj(
+      "draftReturnId.value"        -> draftReturnId.value,
+      "upscanUploadMeta.reference" -> upscanReference.value
+    )
+    EitherT[Future, Error, Option[UpscanUpload]](
+      collection
+        .find(selector, None)
+        .one[UpscanUpload]
+        .map(Right(_))
+        .recover {
+          case exception => Left(Error(exception))
+        }
+    )
+  }
+
+  override def update(
+    draftReturnId: DraftReturnId,
+    upscanReference: UpscanReference,
+    upscanUpload: UpscanUpload
+  ): EitherT[Future, Error, Unit] = {
+
+    val selector = Json.obj(
+      "draftReturnId.value"        -> draftReturnId.value,
+      "upscanUploadMeta.reference" -> upscanReference.value
+    )
+
+    val updatedUpscanUpload = Json.toJsObject(upscanUpload)
+
+    EitherT[Future, Error, Unit](
+      findAndUpdate(
+        selector,
+        updatedUpscanUpload,
+        fetchNewObject = false,
+        upsert         = false
+      ).map {
+        _.lastError.flatMap(_.err) match {
+          case Some(error) => Left(Error(s"could not update upscan upload: $error"))
+          case None        => Right(())
+        }
+      }
+    )
+  }
+
+  override val indexName: String = "upscan-cache-ttl"
+  override val objName: String   = ""
```

Comment:
why do I need to supply this objName ? It complained about it when I added the CacheRepository

Reply:
the `CacheRepository` stores things in mongo like this:
```
{
  "_id": "abc",
   "lastUpdated": "some-date-time",
   "objName": { // whatever you tell it to write into mongo //  }
}
```
so the `objName` controls the path of what you write to mongo. So in your selector above:
```scala
val selector = Json.obj(
      "draftReturnId.value"        -> draftReturnId.value,
      "upscanUploadMeta.reference" -> upscanReference.value
    )
```
you'll have to prefix the things on the right with `s"$key.blah"`


```

Reply:
what would be a good key for this?

Reply:
`upscan` maybe?

Reply:
done